ggshield/config.py

```diff
@@ -1,322 +1,470 @@
 import copy
-import json
 import os
-from typing import Any, Dict, List, NamedTuple, Set
+from dataclasses import dataclass, field, fields, is_dataclass
+from datetime import datetime, timedelta
+from pathlib import Path
+from typing import Any, Dict, Iterable, List, Mapping, Optional, Set, Tuple, Type, Union
 
 import click
 import yaml
-from dotenv import load_dotenv
-from pygitguardian.models import PolicyBreak
+from appdirs import user_config_dir
 
-from ggshield.config_types import IgnoredMatch
-from ggshield.filter import get_ignore_sha
+from ggshield.constants import (
+    AUTH_CONFIG_FILENAME,
+    DEFAULT_LOCAL_CONFIG_PATH,
+    GLOBAL_CONFIG_FILENAMES,
+    LOCAL_CONFIG_PATHS,
+)
+from ggshield.types import IgnoredMatch
+from ggshield.utils import api_to_dashboard_url, dashboard_to_api_url
 
-from .git_shell import get_git_root, is_git_dir
-from .text_utils import display_error
 
-
-CONTEXT_SETTINGS = dict(help_option_names=["-h", "--help"])
-
-# max file size to accept
-MAX_FILE_SIZE = 1048576
-
-CPU_COUNT = os.cpu_count() or 1
-
-
-class Attribute(NamedTuple):
-    name: str
-    default: Any
-
-
-def replace_in_keys(data: Dict, old_char: str, new_char: str) -> None:
+def replace_in_keys(
+    data: Union[List, Dict], old_char: str, new_char: str, recursive: bool = True
+) -> None:
     """Replace old_char with new_char in data keys."""
-    for key in list(data):
-        if old_char in key:
-            new_key = key.replace(old_char, new_char)
-            data[new_key] = data.pop(key)
+    if recursive:
+        if isinstance(data, dict):
+            for key, value in list(data.items()):
```

Comment:
no, because we are modifying the keys during iteration, it causes an error otherwise:

```py
RuntimeError: dictionary keys changed during iteration
```